docs/flask.md

```diff
@@ -0,0 +1,202 @@
+```
+
+To get a Python REPL in the Flask environment for quickly testing an idea, run
+
+```bash
+docker-compose exec app flask shell
+```
+
+If your editor is not configured to run the [Black formatter](https://black.readthedocs.io/)
+automatically on save and you do not have it installed globally or in a virtualenv
+(in which case you can just run `black` on the changed source files), you can
+run the formatter in a Docker container before committing:
+
+```bash
+docker-compose run --rm app black
+```
+
+## [Pytest](https://docs.pytest.org/)
+
+Pytest is used for integration [testing](https://github.com/ccmbioinfo/stager/tree/master/flask/tests)
+the backend. This means that we need a separate test database in MySQL, and a separate testing MinIO
+instance. Docker facilitates spinning these up really conveniently, but you can also manually create
+a new database and a separate MinIO listening on a different port.
+
+All endpoints should ideally have a set of comprehensive tests so we maintain confidence in
+dependency updates, refactors, and new features implemented.
+Each `test_*.py` file in the `tests` directory tests to a set of endpoints implemented by the
+corresponding Blueprint in the `app` directory.
+
+### Setting up
+
+Make sure all the `TEST_*` environment variables in `.env` are set.
+See [`sample.env`](https://github.com/ccmbioinfo/stager/blob/master/sample.env) for examples.
+These are used to initialize a different config object in [`conftest.py`](https://github.com/ccmbioinfo/stager/blob/master/flask/tests/conftest.py).
+
+To start the separate test service containers in the background, run
+
+```bash
+docker-compose -p stager-test -f docker-compose.test.yaml up -d mysql minio
+```
+
+This step is optional if you are running the tests in Docker, as the next
+steps will start these containers anyway.
+
+These containers do not have their ports exposed to the host like the usual
+development containers. To expose them to the host, add the `ports:` key to each
+service like in the the regular `docker-compose.yaml`. This is useful if
+running pytest via PyCharm for debugging.
+
+`-p` is short for `--project-name` and keeps these containers in a separate namespace
+so they don't replace your regular development containers. To run in the foreground,
+remove the `-d` flag.
+
+To run pytest in Docker:
+
+```bash
+docker-compose -p stager-test -f docker-compose.test.yaml run --rm -T app
+```
+
+`--rm` removes the test container after it's done running.
+`-T` is optional. It does not connect a TTY, which used to cause problems with a
+particular test but has since been [resolved](https://github.com/minio/mc/issues/3499).
+
+To run a specific test file or function instead of the entire test suite, append `pytest ARGUMENTS`
+to the command above. For example, `pytest tests/test_users.py` will only run the integration tests
+in the specified file. For more information, see [pytest usage](https://docs.pytest.org/en/stable/usage.html).
```

Comment:
I was actually wondering if there was a way to do this since I've been waiting for all tests to run when adding or testing new tests!